### DISCUSSION TITLE
How to export polynomials from pce in a pre-run

### DISCUSSION THREAD

**Discussion by amorfinv** [snl-dakota/dakota#146] [Q&A] [answered]
Hello,

I am currently trying to get the coefficeints of a Polynomial Chaos Expansion by doing a pre-run. However, I am not getting any data output. I suspect my input file has some missing parameter

```
environment
  tabular_data
    tabular_data_file = 'tabulardata.out'
  pre_run
    output = 'pre.out'

method
  polynomial_chaos
    quadrature_order = 3
    samples = 10000
    num_response_levels = 5661*72
    response_levels = 5661*5:5:360
    variance_based_decomp
    normalized
    export_expansion_file = 'coeff.dat'

variables
  histogram_bin_uncertain = 1
    pairs_per_variable = 73
    abscissas          = 2.5  7.5  12.5  17.5  22.5  27.5  32.5  37.5  42.5  47.5  52.5  57.5  62.5  67.5  72.5  77.5  82.5  87.5  92.5  97.5  102.5  107.5  112.5  117.5  122.5  127.5  132.5  137.5  142.5  147.5  152.5  157.5  162.5  167.5  172.5  177.5  182.5  187.5  192.5  197.5  202.5  207.5  212.5  217.5  222.5  227.5  232.5  237.5  242.5  247.5  252.5  257.5  262.5  267.5  272.5  277.5  282.5  287.5  292.5  297.5  302.5  307.5  312.5  317.5  322.5  327.5  332.5  337.5  342.5  347.5  352.5  357.5  362.5
    counts             = 443  408  421  384  419  469  623  932  1061  1106  980  763  574  519  557  502  575  647  618  619  596  503  408  366  378  343  308  315  383  591  712  906  929  928  822  921  1187  1497  1214  1001  907  1103  1082  1352  1383  1144  1093  1198  1014  928  862  882  916  954  809  789  739  734  595  469  490  437  442  491  424  497  596  639  580  584  630  548  0
    descriptors        = 'DIR'

interface
	fork
	  asynchronous
	  evaluation_concurrency = 3
          local_evaluation_scheduling static
	  analysis_driver = 'simulator_script'
	  parameters_file = 'parameters.in'
	  results_file    = 'results.out'
	  work_directory directory_tag file_tag
	  named 'workdir' file_save  directory_save
 	  link_files = 'templatedir/*'
	  deactivate active_set_vector

responses
  num_response_functions = 5661
  field_responses = 5661
  length = 5661*1
  no_gradients
  no_hessians
```

I thought I would get a file called ```coeff.dat```, but it does not appear. I also tried to add an output to ```pre_run```.

This is the output of dakota

```
---------------------
End DAKOTA input file
---------------------

Using Dakota input file 'pce.in'
Writing new restart file 'dakota.rst'.

>>>>> Executing environment.

>>>>> Running polynomial_chaos iterator.

Pre-run phase complete: no variables to output.


<<<<< Iterator polynomial_chaos completed.
<<<<< Environment execution completed.
DAKOTA execution time in seconds:
  Total CPU        =    0.27265 [parent =   0.272648, child =      2e-06]
  Total wall clock =   0.272679
```

**Comment by jadamstephens** [comment]
Pre-run mode works for a very limited number of methods in Dakota, and it looks like Dakota doesn't provide a meaningful error message when it doesn't work.

Let me make sure I understand. You want Dakota to print out a listing of the parameters it would run to construct the PCE. Then you plan to run your simulation at those points, run Dakota in post-run mode to ingest the responses, construct the PCE, and print out the coefficients?

**Reply by amorfinv** [reply]
Hello, thanks for the response. 

Yes, I am running some CFD simulations at different wind directions. I have the histogram of the wind directions, so I would like Dakota to tell me which wind directions are relevant to run. Then, once I run the CFD simulations outside of Dakota, I would like to feed the results back to Dakota and do the UQ analysis. From my limited understanding of UQ, the PCE coefficients would help me decide which wind directions are most relevant.

In theory, I would prefer to run the CFD simulations within Dakota, but I am working in a cluster with job limits so, it seems easier to run the simulations outside of Dakota.

**Comment by thenectarine** [comment, marked as answer]
This might work for you. The problem is after you run your CFD at the PCE training points, restarting Dakota to do your UQ is an issue.

Manually acquiring training data from CFD

1. I would swap your interface with a dummy interface/function that is cheap to run, see example 1 and 2 below. 
2. Run the PCE to completion with the dummy interface. Since the dummy interface is cheap to run, completion should take a few seconds.
3. The Dakota output file and tabular file, if you have the keyword output_file 'dakota.out' or tabular_data_file 'tabular.data,' would have a listing of all the runs performed. The tabular file is easier to work with, see listing D. If you look at listing A and B, lines 64-136 have the runs and variable values used to construct the PCE. Listing C is the Dakota input file used to generate this output.
4. Run your CFD at those points and collect the responses. This is your training data.

This is where the issue arises. How do you restart Dakota and force Dakota to use your manually acquired training data to build the PCE?

6. Build a new dummy interface/function that only generates responses for the training points. If there were 4 training data points, this function only writes out responses for 4 training data points. Since we know Dakota originally needed N training data points, and the new dummy function provides responses for the N training data points, this might be an acceptable workaround.
7. Use this new dummy function as your interface.
8. Run Dakota to completion.
9. You should have your UQ quantities of interest, e.g. mean, standard deviation and probabilities for your response_levels. In your case, you want the coefficients, so you would see something like this:

```
Coefficients of Polynomial Chaos Expansion for response_fn_1:
        coefficient   u1   u2
        ----------- ---- ----
   1.8020000000e+03  He0  He0
   8.0200000000e+02  He1  He0
  -2.0000000000e+02  He0  He1
  -8.0000000000e+02  He1  He1
```

Example 1
>interface
   analysis_drivers  'rosenbrock'
         direct

Example 2
>interface
   analysis_drivers  'python dummy_blackbox.py'
         fork

Listing A
See file: [dakota.out.txt](https://github.com/user-attachments/files/17421863/dakota.out.txt)

Listing B

> Line    50: Executing environment.
Line    52: Running polynomial_chaos iterator.
Line    54: polynomial_chaos: pre-run phase.
Line    56:  polynomial_chaos: core run phase.
Line    58: Building global_projection_orthogonal_polynomial approximations.
Line    60: Running quadrature iterator.
Line    62: quadrature: pre-run phase.
Line    64: quadrature: core run phase.
Line   129: quadrature: post-run phase.
Line   136: Appending 4 points to global_projection_orthogonal_polynomial approximations.
Line 14146: polynomial_chaos: post-run phase.

Listing C
See file: [study_a.in.txt](https://github.com/user-attachments/files/17421864/study_a.in.txt)

Listind D - tabular.dat - Training points for PCE


```
%eval_id interface x1             x2             response_fn_1  
1        NO_ID      1              3              400            
2        NO_ID      3              3              3604           
3        NO_ID      1              5              1600           
4        NO_ID      3              5              1604           
```

**Reply by amorfinv** [reply]
Hi. Thanks this was quite helpful. I was able to get the relevant training points!

However, had another question about the PCE. I am currently running a ```histogram_bin_uncertain``` case because I have a histogram of the wind directions. It doesn't really fit any of the other distributions.

![figure](https://github.com/user-attachments/assets/3c02d1e9-ec4f-4fd7-9e6c-aea850712865)

I noticed that the training points cannot be higher than 23 for some reason (```quadrature_order=23```).  Also, the training points selected end up being spread in a Gaussian manner (see black dashed lines).

![figure](https://github.com/user-attachments/assets/67e04d34-22c4-4fd7-b09c-6d90a46c7721)


Is there another Coefficient estimation approach recommended when using the ```histogram_bin_uncertain``` variable?



here is my input file for reference:

```
method
  polynomial_chaos
    quadrature_order = 23
    samples = 50000
    seed = 12347
    num_response_levels = 5661*23
    response_levels = 5661*1:1:23
    variance_based_decomp
    normalized
    export_expansion_file = 'coeff1.dat'


variables
  histogram_bin_uncertain = 1
    pairs_per_variable = 50
    abscissas          = 0.0  7.346938775510204  14.693877551020408  22.040816326530614  29.387755102040817  36.734693877551024  44.08163265306123  51.42857142857143  58.775510204081634  66.12244897959184  73.46938775510205  80.81632653061224  88.16326530612245  95.51020408163265  102.85714285714286  110.20408163265306  117.55102040816327  124.89795918367348  132.24489795918367  139.59183673469389  146.9387755102041  154.28571428571428  161.6326530612245  168.9795918367347  176.3265306122449  183.67346938775512  191.0204081632653  198.3673469387755  205.71428571428572  213.06122448979593  220.40816326530611  227.75510204081633  235.10204081632654  242.44897959183675  249.79591836734696  257.14285714285717  264.48979591836735  271.83673469387753  279.18367346938777  286.53061224489795  293.8775510204082  301.2244897959184  308.57142857142856  315.9183673469388  323.265306122449  330.6122448979592  337.9591836734694  345.3061224489796  352.6530612244898  360.0
    counts             = 685  627  575  608  827  1376  1598  1424  984  777  783  816  933  904  866  606  549  519  447  521  900  1210  1334  1299  1401  1957  1841  1378  1564  1765  2003  1620  1749  1475  1261  1310  1418  1147  1153  1057  773  696  631  717  655  833  912  865  890  0
    descriptors        = 'x1'
```
